### DISCUSSION TITLE
Optimization Help

### DISCUSSION THREAD

**Discussion by UndarkAido** [hirrolot/metalang99#21] [Q&A] [answered]
In November I designed https://github.com/DiscordPP/plugin-endpoints/tree/dev/native to use Metalang99 to generate endpoints and objects for the Discord API, but there's a small problem- compilation takes so long that CLion times out CMake, I think while it's doing precompiled headers. Does anyone have any tips for optimizing what I've done, or maybe ways to preprocess things before including them in the project?

**Comment by hirrolot** [comment, marked as answer]
Hello! Have you tried precompiling the headers manually before launching CLion? E.g., there is [`target_precompile_headers`](https://cmake.org/cmake/help/git-stage/command/target_precompile_headers.html) in CMake, which allows to do it automatically. My logic is that after the headers are precompiled, CLion will find the corresponding files in the directory and will not perform the precompiling anymore.

**Reply by hirrolot** [reply]
Also, have you tried asking for help CLion's tech support?

**Reply by UndarkAido** [reply]
No, not yet. Compilation is also slow outside, though.

**Reply by hirrolot** [reply]
Did you specify `-ftrack-macro-expansion=0` for GCC or `-fmacro-backtrace-limit=1` for Clang, as outlined in the README installation section? Probably CLion has settings to specify custom compilation options?

**Reply by UndarkAido** [reply]
Ah, I haven't! I'll give those a try.

**Reply by UndarkAido** [reply]
> Did you specify `-ftrack-macro-expansion=0` for GCC or `-fmacro-backtrace-limit=1` for Clang, as outlined in the README installation section? Probably CLion has settings to specify custom compilation options?

It was this! Thank you so much.

**Comment by UndarkAido** [comment]
I'm not 100% certain but I think that's actually the step where CLion times
out

**Reply by hirrolot** [reply]
Also, you can try precompiling the whole `metalang99/include` directory. There are a lot of macros within, so probably they slow down CLion as well.

**Comment by hirrolot** [comment]
@UndarkAido, does precompiling the headers before launching CLion help? In particular, precompiling the whole `metalang99/include` directory.

**Reply by UndarkAido** [reply]
Yeah- it works, but it's an absolute pain and you have to do it every time you touch CMake
